File: src/ui/pages/history.py
import threading
from typing import Callable

from gi.repository import Gtk, Adw, GLib, GObject, Pango

from api.client import MusicClient
from ui.utils import AsyncPicture, LikeButton
from ui.context_menu import MenuAction, show_song_menu
from ui.util_classes import ScrolledWindow


class HistoryPage(Adw.Bin):
    """Listening-history page, grouped by the `played` field that
    ytmusicapi's get_history() attaches to each track ("Today",
    "Yesterday", "This week", ...).

    Visually modelled on YT Music's Verlauf/History screen:
    section heading + row per track (thumbnail, title, artist, duration),
    no big play-all header. Clicking a row plays it and queues the
    remaining flat history after it.
    """

    __gsignals__ = {
        "header-title-changed": (GObject.SignalFlags.RUN_FIRST, None, (str,)),
    }

    def __init__(self, player, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.player = player
        self.client = MusicClient()
        self._tracks = []
        # Collected on each rebuild so set_compact_mode can swap the
        # thumbnail sizing on every row (matches PlaylistPage).
        self._row_imgs = []
        self._compact = False

        self.main_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)
        self.set_child(self.main_box)

        self.scrolled = ScrolledWindow()
        self.scrolled.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.AUTOMATIC)
        self.scrolled.set_vexpand(True)

        self.clamp = Adw.Clamp()
        self.clamp.set_maximum_size(1024)
        self.clamp.set_tightening_threshold(600)

        self.content_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=24)
        self.content_box.set_margin_top(24)
        self.content_box.set_margin_bottom(24)
        self.content_box.set_margin_start(24)
        self.content_box.set_margin_end(24)

        # Page title ("Listening History")
        self.title_label = Gtk.Label(label="Listening History")
        self.title_label.add_css_class("title-1")
        self.title_label.set_halign(Gtk.Align.START)
        self.content_box.append(self.title_label)

        # Sections container — emptied and rebuilt on every fetch.
        self.sections_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=16)
        self.content_box.append(self.sections_box)

        # Empty-state label (shown when history returns zero tracks).
        self.empty_label = Gtk.Label(
            label="Your listening history will appear here after you play something."
        )
        self.empty_label.add_css_class("dim-label")
        self.empty_label.set_wrap(True)
        self.empty_label.set_halign(Gtk.Align.CENTER)
        self.empty_label.set_margin_top(48)
        self.empty_label.set_visible(False)
        self.content_box.append(self.empty_label)

        self.clamp.set_child(self.content_box)
        self.scrolled.set_child(self.clamp)

        # Loading spinner overlayed at the Bin level — same reliable
        # centering pattern as LibraryPage: Gtk.Overlay sidesteps the
        # Adw.Clamp/ScrolledWindow sizing fight that breaks vexpand
        # chains inside the viewport.
        self._loading_wrap = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=12)
        self._loading_wrap.set_valign(Gtk.Align.CENTER)
        self._loading_wrap.set_halign(Gtk.Align.CENTER)
        self._loading_wrap.set_visible(True)
        spinner = Adw.Spinner()
        spinner.set_size_request(48, 48)
        self._loading_wrap.append(spinner)
        lbl = Gtk.Label(label="Loading history...")
        lbl.add_css_class("caption")
        self._loading_wrap.append(lbl)

        overlay = Gtk.Overlay()
        overlay.set_vexpand(True)
        overlay.set_child(self.scrolled)
        overlay.add_overlay(self._loading_wrap)
        self.main_box.append(overlay)

        # Title-in-headerbar behavior matches PlaylistPage so the
        # NavigationPage's title bar stays in sync when scrolling.
        self.scrolled.get_vadjustment().connect("value-changed", self._on_scroll)

    # ── Loading ────────────────────────────────────────────────────────────

    def load(self):
        """Kick off the initial render.

        Cached tracks are loaded and rendered asynchronously. Once cached rendering completes,
        a refresh callback is triggered to fetch fresh history from the server.
        """
        self._load_cached(async_render=True, async_callback=self.refresh_from_server)

    def _load_cached(self, async_render=False, async_callback: Callable[[], None]=lambda:None):
        if not self.client.is_authenticated():
            self._show_empty("Sign in to view listening history.")
            return
        cached = self.client.get_cached_history() or []
        if cached:
            self._normalize_durations(cached)
            self._render(cached, async_render=async_render, async_callback=async_callback)
        else:
            self._clear_sections()
            self.empty_label.set_visible(False)
            self._loading_wrap.set_visible(True)

    def load_cached(self):
        """Paint the cached history immediately. Intended to run before
        the nav animation so the pushed page isn't blank."""
        self._load_cached()

    def refresh_from_server(self, delay_ms=0):
        """Fetch fresh history and re-render. `delay_ms` lets callers
        push the network work past the nav-transition frames."""
        from ui.utils import is_online
        if not self.client.is_authenticated():
            return
        if not is_online():
            if not self._tracks:
                self._show_empty("History requires an internet connection.")
            return

        def _fetch():
            tracks = self.client.get_history() or []
            self._normalize_durations(tracks)
            GLib.idle_add(self._render, tracks)

        def _kick():
            threading.Thread(target=_fetch, daemon=True).start()
            return False

        if delay_ms > 0:
            GLib.timeout_add(delay_ms, _kick)
        else:
            _kick()

    @staticmethod
    def _normalize_durations(tracks):
        """Fill in `duration_seconds` from the "3:42"-style string that
        ytmusicapi returns. The player queue expects numeric seconds.

        Also derives a clean `artist` string from the `artists` list,
        skipping the view-count entries ytmusicapi tacks on for history
        items ("7.2M views"). If we don't set `artist` here, the player
        falls back to joining every `artists[].name` — producing
        "Jamie Paige, 7.2M views" in the player bar."""
        import re
        view_re = re.compile(
            r"^\s*\d+(?:[.,]\d+)?\s*[kKmMbB]?\s*views?\s*$"
        )
        for t in tracks:
            if not t.get("duration_seconds"):
                dstr = t.get("duration") or ""
                parts = dstr.split(":")
                try:
                    if len(parts) == 2:
                        t["duration_seconds"] = int(parts[0]) * 60 + int(parts[1])
                    elif len(parts) == 3:
                        t["duration_seconds"] = (
                            int(parts[0]) * 3600
                            + int(parts[1]) * 60
                            + int(parts[2])
                        )
                except ValueError:
                    pass

            if not t.get("artist"):
                real_artists = [
                    a.get("name", "")
                    for a in (t.get("artists") or [])
                    if isinstance(a, dict)
                    and a.get("name")
                    and not view_re.match(a["name"])
                ]
                if real_artists:
                    t["artist"] = ", ".join(real_artists)

    def _render(self, tracks, async_render=False, async_callback: Callable[[], None]=lambda:None):
        """Renders the given tracks onto the page.

        If async_render is True, rendering is performed incrementally using GLib.idle_add,
        processing one group per idle cycle. async_callback (if provided) is called once
        rendering completes (when idle returns False).

        If async_render is False, all groups are rendered synchronously in a single pass
        without idle scheduling. async_callback will not be called.
        """
        self._loading_wrap.set_visible(False)
        self._tracks = tracks
        self._row_imgs = []
        self._clear_sections()
        if not tracks:
            self.empty_label.set_label(
                "Your listening history will appear here after you play something."
            )
            self.empty_label.set_visible(True)
            return
        self.empty_label.set_visible(False)

        # Group by the `played` header YT Music attaches to each track.
        # Tracks that somehow lack a value fall back to "Recently".
        by_title = {}
        for idx, t in enumerate(tracks):
            t["_history_index"] = idx
            key = t.get("played") or "Recently"
            if key not in by_title:
                by_title[key] = []
            by_title[key].append(t)

        groups = by_title.items()  # [(title, [tracks])], preserves YT's order

        if async_render:
            groups_iter = iter(groups)

            def _render_by_group():
                item = next(groups_iter, None)
                if item is None:
                    async_callback()
                    return False
                title, group_tracks = item
                self.sections_box.append(self._make_section(title, group_tracks))
                return True

            GLib.idle_add(_render_by_group)
        else:
            for title, group_tracks in groups:
                self.sections_box.append(self._make_section(title, group_tracks))

    def _show_empty(self, msg):
        self._loading_wrap.set_visible(False)
        self._clear_sections()
        self.empty_label.set_label(msg)
        self.empty_label.set_visible(True)

    def _clear_sections(self):
        child = self.sections_box.get_first_child()
        while child:
            nxt = child.get_next_sibling()
            self.sections_box.remove(child)
            child = nxt

    # ── Section / row building ─────────────────────────────────────────────

    def _make_section(self, title, tracks):
        box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=8)

        header = Gtk.Label(label=title)
        header.add_css_class("title-3")
        header.set_halign(Gtk.Align.START)
        box.append(header)

        listbox = Gtk.ListBox()
        listbox.set_selection_mode(Gtk.SelectionMode.NONE)
        listbox.add_css_class("boxed-list")
        for t in tracks:
            row = self._make_row(t)
            if row is not None:
                listbox.append(row)
        listbox.connect("row-activated", self._on_row_activated)
        box.append(listbox)
        return box

    def _make_row(self, track):
        vid = track.get("videoId")
        if not vid:
            return None

        row = Gtk.ListBoxRow()
        row._track = track

        hb = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=12)
        hb.set_hexpand(True)
        hb.add_css_class("song-row")

        # Thumbnail — matches playlist rows (56px via `song-img` class).
        thumb_url = ""
        thumbnails = track.get("thumbnails") or []
        if thumbnails:
            thumb_url = thumbnails[-1].get("url", "")
        img = AsyncPicture(
            crop_to_square=True, target_size=56, player=self.player
        )
        img.add_css_class("song-img")
        if self._compact:
            img.set_compact(True)
        if thumb_url:
            img.video_id = vid
            img.load_url(thumb_url)
        hb.append(img)
        self._row_imgs.append(img)

        # Title + subtitle
        vb = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=2)
        vb.set_valign(Gtk.Align.CENTER)
        vb.set_hexpand(True)

        title = track.get("title", "Unknown")
        title_label = Gtk.Label(label=title)
        title_label.set_halign(Gtk.Align.START)
        title_label.set_ellipsize(Pango.EllipsizeMode.END)
        title_label.set_lines(1)
        title_label.set_tooltip_text(title)

        title_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=6)
        title_box.append(title_label)

        # Explicit badge (matches playlist row).
        is_explicit = bool(track.get("isExplicit"))
        if is_explicit:
            explicit_badge = Gtk.Label(label="E")
            explicit_badge.add_css_class("explicit-badge")
            explicit_badge.set_valign(Gtk.Align.CENTER)
            title_box.append(explicit_badge)

        # Download indicator, same icon/size as PlaylistPage.
        dl_icon = Gtk.Image.new_from_icon_name("folder-download-symbolic")
        dl_icon.set_pixel_size(14)
        dl_icon.add_css_class("dim-label")
        dl_icon.set_valign(Gtk.Align.CENTER)
        dl_icon.set_visible(
            self.player.download_manager.is_downloaded(vid)
        )
        title_box.append(dl_icon)

        vb.append(title_box)

        artists = track.get("artists") or []
        artist_str = ", ".join(
            a.get("name", "") for a in artists if isinstance(a, dict)
        )
        album = track.get("album") or {}
        if isinstance(album, dict) and album.get("name"):
            subtitle = (
                f"{artist_str} • {album['name']}" if artist_str else album["name"]
            )
        else:
            subtitle = artist_str
        sub_label = Gtk.Label(label=subtitle or "")
        sub_label.add_css_class("dim-label")
        sub_label.add_css_class("caption")
        sub_label.set_halign(Gtk.Align.START)
        sub_label.set_ellipsize(Pango.EllipsizeMode.END)
        sub_label.set_lines(1)
        sub_label.set_tooltip_text(subtitle or "")
        vb.append(sub_label)

        hb.append(vb)

        # Duration
        dur = track.get("duration_seconds") or 0
        dur_str = track.get("duration") or (
            f"{dur // 60}:{dur % 60:02d}" if dur else ""
        )
        dur_label = Gtk.Label(label=dur_str)
        dur_label.add_css_class("caption")
        dur_label.set_valign(Gtk.Align.CENTER)
        dur_label.set_margin_end(6)
        hb.append(dur_label)

        # Like button
        like = LikeButton(self.client, None)
        like.set_valign(Gtk.Align.CENTER)
        like.set_data(vid, track.get("likeStatus", "INDIFFERENT"))
        hb.append(like)

        row.set_child(hb)

        # Right-click / long-press context menu
        click = Gtk.GestureClick()
        click.set_button(3)
        click.connect("released", self._on_row_right_click, row)
        row.add_controller(click)
        lp = Gtk.GestureLongPress()
        lp.connect(
            "pressed",
            lambda g, x, y, r=row: self._on_row_right_click(g, 1, x, y, r),
        )
        row.add_controller(lp)

        return row

    # ── Interactions ───────────────────────────────────────────────────────

    def _on_row_activated(self, listbox, row):
        track = getattr(row, "_track", None)
        if not track or not self._tracks:
            return
        start = track.get("_history_index", 0)
        # Play from start index through the rest of the history, same
        # as clicking inside a flat playlist.
        self.player.set_queue(
            self._tracks,
            start,
            shuffle=False,
            source_id="HISTORY",
            is_infinite=False,
        )

    def _on_row_right_click(self, gesture, n_press, x, y, row):
        track = getattr(row, "_track", None)
        if not track:
            return
        vid = track.get("videoId")
        if not vid:
            return

        extras = [
            MenuAction(
                "Play",
                lambda r=row: self._on_row_activated(None, r),
                section="queue",
                first=True,
            )
        ]

        # Remove from history is optimistic: the row disappears instantly
        # and the cache is patched up front, the API call is fire-and-forget.
        fb_token = track.get("feedbackToken") or track.get("feedbackTokens")
        if fb_token:
            tokens = (
                [fb_token]
                if isinstance(fb_token, str)
                else list(fb_token.values())
                if isinstance(fb_token, dict)
                else list(fb_token)
            )
            extras.append(
                MenuAction(
                    "Remove from History",
                    lambda v=vid, t=tokens: self._remove_track_optimistic(v, t),
                    section="remove",
                )
            )

        show_song_menu(
            row,
            x,
            y,
            track,
            player=self.player,
            client=self.client,
            prefix="row",
            extras=extras,
        )

    def _remove_track_optimistic(self, video_id, tokens):
        """Delete a track row in-place without a full re-render: drop it
        from `self._tracks` + the disk cache, remove the widget from its
        section's ListBox (and the section itself if it becomes empty),
        then fire the API call in the background."""
        if not video_id:
            return

        # 1. Trim in-memory and disk state.
        self._tracks = [t for t in self._tracks if t.get("videoId") != video_id]
        try:
            self.client.invalidate_history_cache_entry(video_id)
        except Exception as e:
            print(f"[HISTORY] cache invalidate failed: {e}")

        # 2. Find and remove the widget row. Each section is a VBox with
        # [header_label, listbox]; walk them until we hit the matching vid,
        # then stop — there's only ever one row per videoId.
        section = self.sections_box.get_first_child()
        while section is not None:
            next_section = section.get_next_sibling()
            listbox = section.get_last_child()
            removed = False
            if isinstance(listbox, Gtk.ListBox):
                row = listbox.get_first_child()
                while row is not None:
                    nxt = row.get_next_sibling()
                    if getattr(row, "_track", {}).get("videoId") == video_id:
                        listbox.remove(row)
                        removed = True
                        # If that emptied the section, drop its header too.
                        if listbox.get_first_child() is None:
                            self.sections_box.remove(section)
                        break
                    row = nxt
            if removed:
                break
            section = next_section

        # 3. Empty-state banner if that was the last entry.
        if not self._tracks:
            self.empty_label.set_label(
                "Your listening history will appear here after you play something."
            )
            self.empty_label.set_visible(True)

        # 4. Fire the API call.
        def _th():
            try:
                self.client.remove_history_items(tokens)
            except Exception as e:
                print(f"[HISTORY] remove failed: {e}")

        threading.Thread(target=_th, daemon=True).start()

    # ── Title-in-headerbar ─────────────────────────────────────────────────

    def _on_scroll(self, vadjust):
        val = vadjust.get_value()
        self.emit("header-title-changed", "" if val <= 50 else "Listening History")

    def set_compact_mode(self, compact):
        self._compact = compact
        # Shrink/expand every row's thumbnail the same way PlaylistPage
        # propagates compact state to `_lv_img.set_compact(...)`.
        for img in self._row_imgs:
            if hasattr(img, "set_compact"):
                img.set_compact(compact)
        if compact:
            self.add_css_class("compact")
            self.content_box.set_margin_start(12)
            self.content_box.set_margin_end(12)
        else:
            self.remove_css_class("compact")
            self.content_box.set_margin_start(24)
            self.content_box.set_margin_end(24)

File: rust/src/ui/pages/history.rs
//! Port of ui/pages/history.py: the plays YouTube has recorded, grouped under
//! the headings it filed them under. A row plays from there through the rest
//! of the history, and its menu can forget it.

use std::cell::{Cell, RefCell};
use std::rc::Rc;

use gtk::{glib, prelude::*};

use crate::model::{ItemKind, MediaItem, Track};
use crate::net::history::{self, HistoryEntry};
use crate::ui::context::UiContext;
use crate::ui::context_menu::{MenuAction, Section};
use crate::ui::widgets::song_row::SongRow;

/// How far the page scrolls before the header takes over the title.
const TITLE_HANDOVER: f64 = 50.0;
const EMPTY_TEXT: &str = "Your listening history will appear here after you play something.";
/// The queue a history row plays is the history itself, not a playlist.
const QUEUE_SOURCE: &str = "HISTORY";

type TitleListener = Box<dyn Fn(&str)>;

pub struct HistoryPage {
    root: gtk::Box,
    content_box: gtk::Box,
    sections_box: gtk::Box,
    empty_label: gtk::Label,
    loading_wrap: gtk::Box,
    ctx: Rc<UiContext>,
    /// Every play in one flat list, which is what a row queues.
    entries: RefCell<Vec<HistoryEntry>>,
    rows: RefCell<Vec<Rc<SongRow>>>,
    loading: Cell<bool>,
    on_title: RefCell<Option<TitleListener>>,
}

impl HistoryPage {
    pub fn new(ctx: Rc<UiContext>) -> Rc<Self> {
        let root = gtk::Box::builder().orientation(gtk::Orientation::Vertical).build();
        let scrolled = gtk::ScrolledWindow::builder().hscrollbar_policy(gtk::PolicyType::Never).vscrollbar_policy(gtk::PolicyType::Automatic).vexpand(true).build();
        crate::ui::suppress_hover_while_scrolling(&scrolled);

        let content_box = gtk::Box::builder().orientation(gtk::Orientation::Vertical).spacing(24).margin_top(24).margin_bottom(24).margin_start(24).margin_end(24).build();
        content_box.append(&gtk::Label::builder().label("Listening History").css_classes(["title-1"]).halign(gtk::Align::Start).build());

        let sections_box = gtk::Box::builder().orientation(gtk::Orientation::Vertical).spacing(16).build();
        content_box.append(&sections_box);

        let empty_label = gtk::Label::builder().label(EMPTY_TEXT).css_classes(["dim-label"]).wrap(true).halign(gtk::Align::Center).margin_top(48).visible(false).build();
        content_box.append(&empty_label);

        let clamp = adw::Clamp::builder().maximum_size(1024).tightening_threshold(600).child(&content_box).build();
        scrolled.set_child(Some(&clamp));

        // Overlaid rather than packed: the same centring the library page uses,
        // which the clamp inside a scrolled window would otherwise fight.
        let loading_wrap = gtk::Box::builder().orientation(gtk::Orientation::Vertical).spacing(12).valign(gtk::Align::Center).halign(gtk::Align::Center).build();
        let spinner = adw::Spinner::new();
        spinner.set_size_request(48, 48);
        loading_wrap.append(&spinner);
        loading_wrap.append(&gtk::Label::builder().label("Loading history...").css_classes(["caption"]).build());
        let overlay = gtk::Overlay::builder().vexpand(true).child(&scrolled).build();
        overlay.add_overlay(&loading_wrap);
        root.append(&overlay);

        let page = Rc::new(Self {
            root,
            content_box,
            sections_box,
            empty_label,
            loading_wrap,
            ctx,
            entries: RefCell::new(Vec::new()),
            rows: RefCell::new(Vec::new()),
            loading: Cell::new(false),
            on_title: RefCell::new(None),
        });

        let weak = Rc::downgrade(&page);
        scrolled.vadjustment().connect_value_changed(move |adj| {
            if let Some(p) = weak.upgrade() {
                p.emit_title(if adj.value() > TITLE_HANDOVER { "Listening History" } else { "" });
            }
        });
        let weak = Rc::downgrade(&page);
        page.ctx.on_compact(move |compact| match weak.upgrade() {
            Some(p) => {
                p.set_compact_mode(compact);
                true
            }
            None => false,
        });
        page
    }

    pub fn widget(&self) -> &gtk::Box {
        &self.root
    }

    /// The window sets its title from this, like the header-title-changed signal.
    pub fn set_on_header_title(&self, f: impl Fn(&str) + 'static) {
        self.on_title.replace(Some(Box::new(f)));
    }

    fn emit_title(&self, title: &str) {
        if let Some(f) = self.on_title.borrow().as_ref() {
            f(title);
        }
    }

    pub fn set_compact_mode(&self, compact: bool) {
        if compact {
            self.root.add_css_class("compact");
            self.content_box.set_margin_start(12);
            self.content_box.set_margin_end(12);
        } else {
            self.root.remove_css_class("compact");
            self.content_box.set_margin_start(24);
            self.content_box.set_margin_end(24);
        }
    }

    /// Port of load: the cached plays go up straight away, then the fetch
    /// replaces them. Without a cache the page waits on the spinner.
    pub fn load(self: &Rc<Self>) {
        if !self.ctx.player.state().authenticated() {
            self.show_empty("Sign in to view listening history.");
            return;
        }
        let cached = history::cached_history(self.ctx.downloads.store());
        if !cached.is_empty() {
            self.render(cached);
        }
        self.refresh();
    }

    /// Whether a fetch is in flight, which is what the header's refresh
    /// button waits on before putting its icon back.
    pub fn is_loading(&self) -> bool {
        self.loading.get()
    }

    /// Port of refresh_from_server.
    pub fn refresh(self: &Rc<Self>) {
        if !self.ctx.player.state().authenticated() || self.loading.get() {
            return;
        }
        if !self.ctx.online.is_online() {
            if self.entries.borrow().is_empty() {
                self.show_empty("History requires an internet connection.");
            }
            return;
        }
        self.loading.set(true);
        let api = self.ctx.net.client().api();
        let handle = self.ctx.net.spawn(history::get_history(api));
        let weak = Rc::downgrade(self);
        glib::spawn_future_local(async move {
            let outcome = handle.await;
            let Some(page) = weak.upgrade() else { return };
            page.loading.set(false);
            match outcome {
                Ok(Ok(entries)) => {
                    history::cache_history(page.ctx.downloads.store(), &entries);
                    page.render(entries);
                }
                Ok(Err(err)) => {
                    tracing::warn!(%err, "history fetch failed");
                    if page.entries.borrow().is_empty() {
                        page.show_empty(EMPTY_TEXT);
                    }
                }
                Err(_) => {}
            }
        });
    }

    fn render(self: &Rc<Self>, entries: Vec<HistoryEntry>) {
        self.loading_wrap.set_visible(false);
        clear_children(&self.sections_box);
        self.rows.borrow_mut().clear();
        if entries.is_empty() {
            self.entries.borrow_mut().clear();
            self.show_empty(EMPTY_TEXT);
            return;
        }
        self.empty_label.set_visible(false);

        // One section per heading, in the order YouTube sent them.
        let mut sections: Vec<(String, Vec<usize>)> = Vec::new();
        for (index, entry) in entries.iter().enumerate() {
            match sections.last_mut() {
                Some((title, rows)) if *title == entry.played => rows.push(index),
                _ => sections.push((entry.played.clone(), vec![index])),
            }
        }
        for (title, indexes) in sections {
            self.sections_box.append(&self.build_section(&title, &indexes, &entries));
        }
        self.entries.replace(entries);
    }

    fn build_section(self: &Rc<Self>, title: &str, indexes: &[usize], entries: &[HistoryEntry]) -> gtk::Box {
        let section = gtk::Box::builder().orientation(gtk::Orientation::Vertical).spacing(8).build();
        section.append(&gtk::Label::builder().label(title).css_classes(["title-3"]).halign(gtk::Align::Start).build());
        let list = gtk::ListBox::builder().css_classes(["boxed-list", "songs-list"]).selection_mode(gtk::SelectionMode::None).build();

        for &index in indexes {
            let entry = &entries[index];
            let item = as_item(&entry.track);
            let row = SongRow::new(self.ctx.clone());
            row.set_plain_subtitle(Some(subtitle(&entry.track)));
            row.bind(&item, None);
            let weak = Rc::downgrade(self);
            row.set_on_activate(move |_| {
                if let Some(page) = weak.upgrade() {
                    page.play_from(index);
                }
            });
            let weak = Rc::downgrade(self);
            row.set_menu_extras(move || weak.upgrade().map(|page| page.row_menu_extras(index)).unwrap_or_default());
            list.append(row.widget());
            self.rows.borrow_mut().push(row);
        }

        let weak = Rc::downgrade(self);
        let indexes = indexes.to_vec();
        list.connect_row_activated(move |_, row| {
            if let (Some(page), Some(index)) = (weak.upgrade(), indexes.get(row.index().max(0) as usize)) {
                page.play_from(*index);
            }
        });
        section.append(&list);
        section
    }

    /// Port of _on_row_activated: play from here through the rest of the
    /// history, the way a click inside a flat playlist does.
    fn play_from(&self, index: usize) {
        let tracks: Vec<Track> = self.entries.borrow().iter().map(|entry| entry.track.clone()).collect();
        if index >= tracks.len() {
            return;
        }
        self.ctx.player.play_tracks(tracks, index, false, Some(QUEUE_SOURCE.to_owned()), false);
    }

    /// The two entries history.py adds to a row's song menu.
    fn row_menu_extras(self: &Rc<Self>, index: usize) -> Vec<MenuAction> {
        let token = self.entries.borrow().get(index).and_then(|entry| entry.feedback_token.clone());
        let video_id = self.entries.borrow().get(index).map(|entry| entry.track.video_id.0.clone()).unwrap_or_default();

        let weak = Rc::downgrade(self);
        let mut extras = vec![MenuAction { first: true, ..MenuAction::new("Play", Section::Queue, move || {
            if let Some(page) = weak.upgrade() {
                page.play_from(index);
            }
        })}];
        // No token means a brand account, where YouTube offers no removal.
        if let Some(token) = token {
            let weak = Rc::downgrade(self);
            extras.push(MenuAction::new("Remove from History", Section::Remove, move || {
                if let Some(page) = weak.upgrade() {
                    page.remove_play(&video_id, &token);
                }
            }));
        }
        extras
    }

    /// Demo hook: what the first row's menu adds on top of the song menu.
    pub fn menu_extras_for_demo(self: &Rc<Self>) -> Vec<String> {
        if self.entries.borrow().is_empty() {
            return Vec::new();
        }
        self.row_menu_extras(0).into_iter().map(|action| action.label).collect()
    }

    /// Port of _remove_track_optimistic: the row goes now, the cache is
    /// patched now, and the account is told afterwards.
    fn remove_play(self: &Rc<Self>, video_id: &str, token: &str) {
        self.entries.borrow_mut().retain(|entry| entry.track.video_id.0 != video_id);
        history::forget_cached(self.ctx.downloads.store(), video_id);
        let remaining = self.entries.borrow().clone();
        self.render(remaining);

        let api = self.ctx.net.client().api();
        let token = token.to_owned();
        self.ctx.net.spawn(async move {
            if let Err(err) = history::remove_history_items(&api, vec![token]).await {
                tracing::warn!(%err, "removing a play failed");
            }
        });
    }

    fn show_empty(&self, message: &str) {
        self.loading_wrap.set_visible(false);
        clear_children(&self.sections_box);
        self.rows.borrow_mut().clear();
        self.empty_label.set_label(message);
        self.empty_label.set_visible(true);
    }
}

/// Port of the row's subtitle: the artists, then the album behind a bullet.
fn subtitle(track: &Track) -> String {
    let album = track.album.as_ref().map(|a| a.name.as_str()).unwrap_or_default();
    match (track.artist.is_empty(), album.is_empty()) {
        (_, true) => track.artist.clone(),
        (true, false) => album.to_owned(),
        (false, false) => format!("{} \u{2022} {album}", track.artist),
    }
}

/// A history row as the boxed-list row wants it.
fn as_item(track: &Track) -> MediaItem {
    MediaItem {
        kind: if track.video_type.as_deref().is_some_and(|t| t != "MUSIC_VIDEO_TYPE_ATV") { ItemKind::Video } else { ItemKind::Song },
        id: track.video_id.0.clone(),
        title: track.title.clone(),
        artists: track.artists.clone(),
        album: track.album.clone(),
        thumb: track.thumb.clone(),
        duration_seconds: track.duration_seconds,
        explicit: track.is_explicit,
        like_status: Some(track.like_status),
        ..MediaItem::default()
    }
}

fn clear_children(container: &gtk::Box) {
    while let Some(child) = container.first_child() {
        container.remove(&child);
    }
}
